### PR TITLE
Add result count to search history

### PR DESCRIPTION
**Pull request by vdavid**
Fixes https://github.com/sourcegraph/sourcegraph/issues/43422

Changes:
 - Added the search query to the `SearchResultsFetched` event log item.
 - Getting the search history data from the `SearchResultsFetched` items rather than `SearchResultsQueried`
 - Calling `addRecentSearch` when fetching the results in `StreamingSearchResults` rather than in the search box.
 - Added `resultCount` and `limitHit` to the `RecentSearch` interface
 - Updated the UI to add the search count data

This is how it looks:

New search history UI

Collateral fix:
- The field `code_search.results.result_count` in the `SearchResultsFetched` event log item [contained the wrong value](https://github.com/sourcegraph/sourcegraph/pull/44898/files#diff-bcede140966fa65d87915c2ff40126d4a2f79ce28f3a689fb6a888026dec5840L179): it contained the number of results that arrived from the streaming search in the first response, which is often way less than the total number of results (18 instead of 500). This now contains the real number of results.

## Test plan

- Tested if the items are displayed correctly
- Tested if new items are being added after each search.
- Tested if legacy items (not containing the two new fields) work correctly.
- Tested it with a screen reader: **it's not accessible:** the screen reader only says "list containing 20 items" but doesn't read the selected items. However, this was a pre-existing condition to my change, so this is a wontfix for this PR.

## App preview:

- [Web](https://sg-web-dv-add-result-count-to-search.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by vdavid** [comment]
@limitedmage @quinnkeast I created this as a first draft. I welcome any feedback on whether this is a direction we want to pursue. I'm not an expert on our search UI, so I definitely didn't want to put any effort into this before getting feedback.

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.28 kb) | -0.00% (-0.01 kb) | -0.00% (-0.29 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/dv/add-result-count-to-search-history/compare-report.html;tab=live_object) for a full comparison between the commits 9c3e7d1 and 00167a9 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by quinnkeast** [comment]
A better visual approach may be to treat the number of results and the time as a single unit, all aligned to the right, e.g.:

`374 results • 3 minutes ago`

This would also make it more explicitly related, aka "running this search 3 minutes ago _returned_ 374 results" rather than "running this search again _will return_ 374 results."

Other other catch would be to handle the string plurality, like:

- No results
- 1 result
- X results
- XXX+ results

I'd suggest minimizing the investment into this, because the present implementation of recent searches may change as we continue to explore [RFC 760](https://docs.google.com/document/d/1eeb63tAEJ3hkJLh5_tTbwlUlfpnP8onJaBqi5_CWrAA/edit#heading=h.trqab8y0kufp).

**Comment by vdavid** [comment]
@quinnkeast I love that approach for the UI. Makes sense, and makes rendering easier, too.
Also thanks for the suggestion of not investing too much time in this! I read the RFC, and I know we plan major changes to the search UI. I'll definitely let this go if I see that it'd need too much effort.

@limitedmage, do you have any experience with [this React error](https://share.cleanshot.com/bnrCUu)? I've been digging and trying, but no luck so far and don't want to waste time on this in case this is something you meet regularly when working on search.

**Comment by limitedmage** [comment]
@vdavid Not super familiar with that error, are you still getting it?

**Comment by limitedmage** [comment]
BTW we do have `pluralize` to handle plurality, [example here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/repo/stats/RepositoryStatsContributorsPage.tsx?L105:61&subtree=true).

**Comment by vdavid** [comment]
> @limitedmage: Not super familiar with that error, are you still getting it?

Yes. 😕 The problem seems to be that in [this line](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@dv/add-result-count-to-search-history/-/blob/client/web/src/search/results/StreamingSearchResults.tsx?L191), `addRecentSearch` modifies state and triggers a refresh in `SearchNavbarItem`. What I don't understand is why that is a problem because `addRecentSearch` is _not_ directly in the render function of `StreamingSearchResults` but inside a `useEffect` callback. I appreciate any ideas. I'll also try #dev-frontend.

**Comment by limitedmage** [comment]
@vdavid Try using `useLayoutEffect` instead of `useEffect`. If that doesn't work, put the line inside a timeout:
```ts
setTimeout(() => addRecentSearch(submittedURLQuery, results.progress.matchCount), 0)
```

**Comment by vdavid** [comment]
Thanks, @limitedmage, the `setTimeout(..., 0)` wrapper did the trick! I've added a comment near it to explain why it's there.

File: client/search-ui/src/input/SearchHistoryDropdown.tsx [hunk 1]
```diff
@@ -149,7 +151,12 @@ const SearchHistoryEntries: React.FunctionComponent<SearchHistoryEntriesProps> =
             onClick={clickHandler}
         >
             {recentSearches.map((search, index) => (
-                <SearchHistoryEntry key={index} index={index} search={search} selected={index === selectedIndex} />
+                <SearchHistoryEntry
+                    key={`${search.timestamp}-${search.query}`}
```
Comment (vdavid) [inline client/search-ui/src/input/SearchHistoryDropdown.tsx:155]: React nagged me that using he array index was not right so I started using a different unique value

File: client/web/src/search/input/useRecentSearches.test.tsx [hunk 2]
```diff
@@ -163,7 +165,7 @@ describe('recentSearches', () => {
             `)
         })
 
-        test('adding an exisitng item to recent searches deduplicates it and puts it at the top', async () => {
+        test('adding an existing item to recent searches de-duplicates it and puts it at the top', async () => {
```
Comment (vdavid) [inline client/web/src/search/input/useRecentSearches.test.tsx:168]: Unrelated: Typo fixes

File: client/search-ui/src/input/SearchHistoryDropdown.tsx [hunk 3]
```diff
@@ -42,6 +43,7 @@ interface SearchHistoryDropdownProps extends TelemetryProps {
 // button and the content
 const popoverPadding = createRectangle(0, 0, 0, 2)
 
+// eslint-disable-next-line react/display-name
 export const SearchHistoryDropdown: React.FunctionComponent<SearchHistoryDropdownProps> = React.memo(
     ({ recentSearches = [], onSelect, className, telemetryService }) => {
```
Comment (limitedmage) [inline client/search-ui/src/input/SearchHistoryDropdown.tsx:48]: You can remove the eslint disable by adding a name to the function: 
```suggestion
export const SearchHistoryDropdown: React.FunctionComponent<SearchHistoryDropdownProps> = React.memo(
    function SearchHistoryDropdown({ recentSearches = [], onSelect, className, telemetryService }) {
```

File: client/web/src/search/results/StreamingSearchResults.tsx [hunk 4]
```diff
@@ -175,8 +177,12 @@ export const StreamingSearchResults: FC<StreamingSearchResultsProps> = props =>
...
+                    query_data: {
+                        combined: submittedURLQuery,
+                    },
                     results: {
-                        results_count: results.results.length,
+                        results_count: results.progress.matchCount,
+                        limit_hit: limitHit(results.progress),
                         any_cloning: results.progress.skipped.some(skipped => skipped.reason === 'repository-cloning'),
                         alert: results.alert ? results.alert.title : null,
                     },
```
Comment (limitedmage) [inline client/web/src/search/results/StreamingSearchResults.tsx:188]: Are these changes required?

File: client/web/src/search/results/StreamingSearchResults.tsx [hunk 5]
```diff
@@ -185,12 +191,18 @@ export const StreamingSearchResults: FC<StreamingSearchResultsProps> = props =>
             if (results.results.length > 0) {
                 telemetryService.log('SearchResultsNonEmpty')
             }
+
+            // setTimeout avoids a "Cannot update a component while rendering a different component" React error
+            setTimeout(
+                () => addRecentSearch(submittedURLQuery, results.progress.matchCount, limitHit(results.progress)),
+                0
+            )
```
Comment (limitedmage) [inline client/web/src/search/results/StreamingSearchResults.tsx:199]: I suggest putting this in its own `useEffect` to keep this one just for logging.

File: client/web/src/search/results/StreamingSearchResults.tsx [hunk 6]
```diff
@@ -185,12 +191,18 @@ export const StreamingSearchResults: FC<StreamingSearchResultsProps> = props =>
             if (results.results.length > 0) {
                 telemetryService.log('SearchResultsNonEmpty')
             }
+
+            // setTimeout avoids a "Cannot update a component while rendering a different component" React error
```
Comment (limitedmage) [inline client/web/src/search/results/StreamingSearchResults.tsx:195]: nit: slight rewording of the comment
```suggestion
            // Add the recent search in the next queue execution to avoid
            // updating a React component while rendering another component.
```

**Review by limitedmage — COMMENTED** [review]
Looks good overall! Left some comments and suggestions.

**Comment by limitedmage** [comment]
Also, can you fix the alignment issues that are more pronounced now with this text? Adding `align-items: baseline;` after [this line](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/search-ui/src/input/SearchHistoryDropdown.module.scss?L14) should do it.

File: client/search-ui/src/input/SearchHistoryDropdown.tsx (hunk 3 above)
Comment (vdavid) [reply client/search-ui/src/input/SearchHistoryDropdown.tsx:48]: Ah, makes sense, this should help with debugging later, thank you!

**Comment by vdavid** [comment]
> Also, can you fix the alignment issues that are more pronounced now with this text? Adding align-items: baseline; after [this line](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/search-ui/src/input/SearchHistoryDropdown.module.scss?L14) should do it.

Wow, good eyes! I could not see what you meant and even after adding the baseline alignment, I could hardly see any difference at first, but then I could see it and it indeed looks nicer, thank you!

**Comment by vdavid** [comment]
@limitedmage, I've done all the requested changes, could you take another look?

File: client/web/src/search/results/StreamingSearchResults.tsx (hunk 4 above)
Comment (limitedmage) [reply client/web/src/search/results/StreamingSearchResults.tsx:188]: I think we only query `SearchResultsFetched` if users don't have a search history (yet) because they used Sourcegraph before we launched the search history dropdown. If my thinking here is correct, these newly logged events will never be used for the search dropdown, and they are not required for this feature. But it might be nice to have it for actual logging purposes.

Comment (vdavid) [reply client/web/src/search/results/StreamingSearchResults.tsx:188]: Oh, I understand now, makes sense!
I assumed that the `event_log` table was the source of truth for recent searches and the TempSettings storage was unreliable / more like a cache! (it has "temp" in its name, after all)
But based on your message, it seems like the source of truth is the TempSettings, and we only rely on `event_log` for legacy situations.
In that case, yeah, this is not needed, but as you said, it might come in handy, so I'll leave it.
